File: Strings_and_String-Manipulation/firstUniqueCharacter_method1_leetcode.py
# problem: https://leetcode.com/problems/first-unique-character-in-a-string/
# - TIME:  O(n)
# - SPACE: O(1) - chars list, which will not scale with input at all.
    # - List/Array solution. (optimal)

class Solution:
    def firstUniqChar(self, s: str) -> int:
        
        chars = [0] * 26                    # use array[26] for keeping char # of occurences of all alpha-characters.
                                            # a=0, b=1, c=2, ... , x=23, y=24, z=25
        
        for i in range(len(s)):             # traverse string once: use ascii value to store char occurences in index corresponding to letter
            chars[ord(s[i])-ord('a')] +=1
        
        for i in range(len(s)):             # traverse string again: use chars # of occurences. first one to have only 1 is firstUniqueCharacter. return i.
            if chars[ord(s[i])-ord('a')] == 1:
                return i
        
        return -1                           # if no char # of occurences = 1, it doesn't exist.
            
        
        

File: Strings_and_String-Manipulation/firstUniqueCharacter_method1_leetcode.java
// problem: https://leetcode.com/problems/first-unique-character-in-a-string/
// - TIME: O(2n) -> O(n)
// - SPACE: O(n)
// - most optimal solution, except utilizing a hashmap.

class Solution {
  public int firstUniqChar(String s) {
      
      // utilize hashmap to keep count of characters
      Map<Character, Integer> charCount = new HashMap<>();
      for(int i=0; i<s.length(); i++){
          char c = s.charAt(i);
          if(charCount.containsKey(c))
              charCount.put(c, charCount.get(c)+1);
          else
              charCount.put(c, 1);
      }
      
      // now go through original string 1 more time, but if current character's count in hashmap is 1, then return that character.
      for(int i=0; i<s.length(); i++){
          char c = s.charAt(i);
          if(charCount.get(c)==1)
              return i;
      }
      
      return -1;
  }
}